File: taurus-protect-sdk-typescript/tests/integration/extended-services.test.ts
/**
 * Integration tests for extended services that previously lacked test coverage.
 *
 * Tests read-only operations for: BusinessRuleService, WebhookService,
 * WebhookCallService, StakingService, FeePayerService, ExchangeService,
 * AssetService, FiatService, JobService, ContractWhitelistingService,
 * and TaurusNetwork (participants, pledges, lending, settlements, sharing).
 *
 * These tests require a live API connection. Configure via environment variables
 * or hard-coded defaults in config.ts.
 */

import { describeIntegration, getTestClient } from "./helpers";

// =============================================================================
// BusinessRuleService
// =============================================================================

describeIntegration("Integration: BusinessRuleService", () => {
  it("should list business rules", async () => {
    const client = getTestClient();
    try {
      const result = await client.businessRules.list();

      expect(result).toBeDefined();
      console.log(`Found ${result.rules.length} business rules`);
      for (const rule of result.rules.slice(0, 5)) {
        console.log(`  Rule: ID=${rule.id}`);
      }
    } finally {
      client.close();
    }
  }, 30000);
});

// =============================================================================
// WebhookService
// =============================================================================

describeIntegration("Integration: WebhookService", () => {
  it("should list webhooks", async () => {
    const client = getTestClient();
    try {
      const webhooks = await client.webhooks.list();

      expect(webhooks).toBeDefined();
      expect(Array.isArray(webhooks)).toBe(true);
      console.log(`Found ${webhooks.length} webhooks`);
      for (const wh of webhooks.slice(0, 5)) {
        console.log(`  Webhook: ID=${wh.id}, URL=${wh.url}`);
      }
    } finally {
      client.close();
    }
  }, 30000);
});

// =============================================================================
// WebhookCallService
// =============================================================================

describeIntegration("Integration: WebhookCallService", () => {
  it("should list webhook calls", async () => {
    const client = getTestClient();
    try {
      const result = await client.webhookCalls.list();

      expect(result).toBeDefined();
      console.log(`Found ${result.calls.length} webhook calls`);
    } finally {
      client.close();
    }
  }, 30000);
});

// =============================================================================
// StakingService
// =============================================================================

describeIntegration("Integration: StakingService", () => {
  it("should get stake accounts", async () => {
    const client = getTestClient();
    try {
      // Use getStakeAccounts which doesn't require specific validator IDs
      const result = await client.staking.getStakeAccounts({
        addressId: "1",
        accountType: "StakeAccountTypeSolana",
      });

      expect(result).toBeDefined();
      console.log("Staking stake accounts response received");
    } catch (error: unknown) {
      // Staking may not be available in all environments
      console.log(
        `Staking not available: ${error instanceof Error ? error.message : String(error)}`
      );
    } finally {
      client.close();
    }
  }, 30000);
});

// =============================================================================
// FeePayerService
// =============================================================================

describeIntegration("Integration: FeePayerService", () => {
  it("should list fee payers", async () => {
    const client = getTestClient();
    try {
      const feePayers = await client.feePayers.list();

      expect(feePayers).toBeDefined();
      expect(Array.isArray(feePayers)).toBe(true);
      console.log(`Found ${feePayers.length} fee payers`);
      for (const fp of feePayers.slice(0, 5)) {
        console.log(`  FeePayer: ID=${fp.id}`);
      }
    } finally {
      client.close();
    }
  }, 30000);
});

// =============================================================================
// ExchangeService
// =============================================================================

describeIntegration("Integration: ExchangeService", () => {
  it("should list exchanges", async () => {
    const client = getTestClient();
    try {
      const result = await client.exchanges.list();

      expect(result).toBeDefined();
      console.log(`Found ${result.items.length} exchanges`);
      for (const ex of result.items.slice(0, 5)) {
        console.log(`  Exchange: ID=${ex.id}`);
      }
    } finally {
      client.close();
    }
  }, 30000);
});

// =============================================================================
// FiatService
// =============================================================================

describeIntegration("Integration: FiatService", () => {
  it("should list fiat providers", async () => {
    const client = getTestClient();
    try {
      const providers = await client.fiatAccounts.getFiatProviders();

      expect(providers).toBeDefined();
      expect(Array.isArray(providers)).toBe(true);
      console.log(`Found ${providers.length} fiat providers`);
      for (const p of providers.slice(0, 5)) {
        console.log(`  Provider: ${p.provider}`);
      }
    } finally {
      client.close();
    }
  }, 30000);
});

// =============================================================================
// JobService
// =============================================================================

describeIntegration("Integration: JobService", () => {
  it("should list jobs", async () => {
    const client = getTestClient();
    try {
      const jobs = await client.jobs.list();

      expect(jobs).toBeDefined();
      expect(Array.isArray(jobs)).toBe(true);
      console.log(`Found ${jobs.length} jobs`);
      for (const job of jobs.slice(0, 5)) {
        console.log(`  Job: Name=${job.name}`);
      }
    } catch (error: unknown) {
      console.log(
        `Jobs not available: ${error instanceof Error ? error.message : String(error)}`
      );
    } finally {
      client.close();
    }
  }, 30000);
});

// =============================================================================
// TaurusNetwork - Participants
// =============================================================================

describeIntegration("Integration: TaurusNetwork Participants", () => {
  it("should list participants", async () => {
    const client = getTestClient();
    try {
      const participants = await client.taurusNetwork.participants.list();

      expect(participants).toBeDefined();
      console.log(`Found ${participants.length} TaurusNetwork participants`);
      for (const p of participants.slice(0, 5)) {
        console.log(`  Participant: ID=${p.id}, Name=${p.name}`);
      }
    } catch (error: unknown) {
      // TaurusNetwork may not be enabled in all environments
      console.log(
        `TaurusNetwork participants not available: ${error instanceof Error ? error.message : String(error)}`
      );
    } finally {
      client.close();
    }
  }, 30000);
});

// =============================================================================
// TaurusNetwork - Pledges
// =============================================================================

describeIntegration("Integration: TaurusNetwork Pledges", () => {
  it("should list pledges", async () => {
    const client = getTestClient();
    try {
      const result = await client.taurusNetwork.pledges.list();

      expect(result).toBeDefined();
      console.log(`Found ${result.pledges.length} TaurusNetwork pledges`);
      for (const pledge of result.pledges.slice(0, 5)) {
        console.log(`  Pledge: ID=${pledge.id}, Status=${pledge.status}`);
      }
    } catch (error: unknown) {
      // TaurusNetwork may not be enabled in all environments
      console.log(
        `TaurusNetwork pledges not available: ${error instanceof Error ? error.message : String(error)}`
      );
    } finally {
      client.close();
    }
  }, 30000);
});

// =============================================================================
// TaurusNetwork - Lending
// =============================================================================

describeIntegration("Integration: TaurusNetwork Lending", () => {
  it("should list lending offers", async () => {
    const client = getTestClient();
    try {
      const { offers } = await client.taurusNetwork.lending.listLendingOffers();

      expect(offers).toBeDefined();
      console.log(`Found ${offers.length} TaurusNetwork lending offers`);
      for (const offer of offers.slice(0, 5)) {
        console.log(`  Offer: ID=${offer.id}`);
      }
    } catch (error: unknown) {
      console.log(
        `TaurusNetwork lending offers not available: ${error instanceof Error ? error.message : String(error)}`
      );
    } finally {
      client.close();
    }
  }, 30000);

  it("should list lending agreements", async () => {
    const client = getTestClient();
    try {
      const { agreements } =
        await client.taurusNetwork.lending.listLendingAgreements();

      expect(agreements).toBeDefined();
      console.log(`Found ${agreements.length} TaurusNetwork lending agreements`);
      for (const agreement of agreements.slice(0, 5)) {
        console.log(`  Agreement: ID=${agreement.id}`);
      }
    } catch (error: unknown) {
      console.log(
        `TaurusNetwork lending agreements not available: ${error instanceof Error ? error.message : String(error)}`
      );
    } finally {
      client.close();
    }
  }, 30000);
});

// =============================================================================
// TaurusNetwork - Settlements
// =============================================================================

describeIntegration("Integration: TaurusNetwork Settlements", () => {
  it("should list settlements", async () => {
    const client = getTestClient();
    try {
      const { settlements } = await client.taurusNetwork.settlements.list();

      expect(settlements).toBeDefined();
      console.log(`Found ${settlements.length} TaurusNetwork settlements`);
      for (const s of settlements.slice(0, 5)) {
        console.log(`  Settlement: ID=${s.id}, Status=${s.status}`);
      }
    } catch (error: unknown) {
      console.log(
        `TaurusNetwork settlements not available: ${error instanceof Error ? error.message : String(error)}`
      );
    } finally {
      client.close();
    }
  }, 30000);
});

// =============================================================================
// TaurusNetwork - Sharing
// =============================================================================

describeIntegration("Integration: TaurusNetwork Sharing", () => {
  it("should list shared addresses", async () => {
    const client = getTestClient();
    try {
      const { sharedAddresses } =
        await client.taurusNetwork.sharing.listSharedAddresses();

      expect(sharedAddresses).toBeDefined();
      console.log(
        `Found ${sharedAddresses.length} TaurusNetwork shared addresses`
      );
      for (const addr of sharedAddresses.slice(0, 5)) {
        console.log(`  SharedAddress: ID=${addr.id}`);
      }
    } catch (error: unknown) {
      console.log(
        `TaurusNetwork shared addresses not available: ${error instanceof Error ? error.message : String(error)}`
      );
    } finally {
      client.close();
    }
  }, 30000);

  it("should list shared assets", async () => {
    const client = getTestClient();
    try {
      const { sharedAssets } =
        await client.taurusNetwork.sharing.listSharedAssets();

      expect(sharedAssets).toBeDefined();
      console.log(`Found ${sharedAssets.length} TaurusNetwork shared assets`);
      for (const asset of sharedAssets.slice(0, 5)) {
        console.log(`  SharedAsset: ID=${asset.id}`);
      }
    } catch (error: unknown) {
      console.log(
        `TaurusNetwork shared assets not available: ${error instanceof Error ? error.message : String(error)}`
      );
    } finally {
      client.close();
    }
  }, 30000);
});

// =============================================================================
// ContractWhitelistingService
// =============================================================================

describeIntegration("Integration: ContractWhitelistingService", () => {
  it("should list whitelisted contracts", async () => {
    const client = getTestClient();
    try {
      const result = await client.contractWhitelisting.list();

      expect(result).toBeDefined();
      console.log(`Found ${result.items.length} whitelisted contracts`);
      for (const c of result.items.slice(0, 5)) {
        console.log(`  Contract: ID=${c.id}`);
      }
    } finally {
      client.close();
    }
  }, 30000);
});

File: taurus-protect-sdk-go/test/integration/extended_services_test.go
package integration

import (
	"context"
	"testing"
)

// =============================================================================
// BusinessRuleService
// =============================================================================

func TestIntegration_ListBusinessRules(t *testing.T) {
	skipIfNotIntegration(t)
	client := getTestClient(t)
	defer client.Close()

	ctx := context.Background()
	result, err := client.BusinessRules().ListBusinessRules(ctx, nil)
	if err != nil {
		t.Fatalf("ListBusinessRules() error = %v", err)
	}

	t.Logf("Found %d business rules", len(result.BusinessRules))
	for _, rule := range result.BusinessRules {
		t.Logf("  Rule: ID=%s", rule.ID)
	}
}

// =============================================================================
// WebhookService
// =============================================================================

func TestIntegration_ListWebhooks(t *testing.T) {
	skipIfNotIntegration(t)
	client := getTestClient(t)
	defer client.Close()

	ctx := context.Background()
	result, err := client.Webhooks().ListWebhooks(ctx, nil)
	if err != nil {
		t.Fatalf("ListWebhooks() error = %v", err)
	}

	t.Logf("Found %d webhooks", len(result.Webhooks))
	for _, wh := range result.Webhooks {
		t.Logf("  Webhook: ID=%s, URL=%s", wh.ID, wh.URL)
	}
}

// =============================================================================
// WebhookCallService
// =============================================================================

func TestIntegration_ListWebhookCalls(t *testing.T) {
	skipIfNotIntegration(t)
	client := getTestClient(t)
	defer client.Close()

	ctx := context.Background()
	result, err := client.WebhookCalls().ListWebhookCalls(ctx, nil)
	if err != nil {
		t.Fatalf("ListWebhookCalls() error = %v", err)
	}

	t.Logf("Found %d webhook calls", len(result.WebhookCalls))
}

// =============================================================================
// FeePayerService
// =============================================================================

func TestIntegration_ListFeePayers(t *testing.T) {
	skipIfNotIntegration(t)
	client := getTestClient(t)
	defer client.Close()

	ctx := context.Background()
	result, err := client.FeePayers().ListFeePayers(ctx, nil)
	if err != nil {
		t.Fatalf("ListFeePayers() error = %v", err)
	}

	t.Logf("Found %d fee payers", len(result.FeePayers))
	for _, fp := range result.FeePayers {
		t.Logf("  FeePayer: ID=%s", fp.ID)
	}
}

// =============================================================================
// ExchangeService
// =============================================================================

func TestIntegration_ListExchangeCounterparties(t *testing.T) {
	skipIfNotIntegration(t)
	client := getTestClient(t)
	defer client.Close()

	ctx := context.Background()
	result, err := client.Exchanges().ListExchangeCounterparties(ctx)
	if err != nil {
		t.Fatalf("ListExchangeCounterparties() error = %v", err)
	}

	t.Logf("Found %d exchange counterparties", len(result.Exchanges))
	for _, cp := range result.Exchanges {
		t.Logf("  Exchange: %s", cp.Name)
	}
}

// =============================================================================
// FiatService
// =============================================================================

func TestIntegration_ListFiatProviders(t *testing.T) {
	skipIfNotIntegration(t)
	client := getTestClient(t)
	defer client.Close()

	ctx := context.Background()
	result, err := client.Fiat().ListFiatProviders(ctx)
	if err != nil {
		t.Fatalf("ListFiatProviders() error = %v", err)
	}

	t.Logf("Found %d fiat providers", len(result.FiatProviders))
	for _, p := range result.FiatProviders {
		t.Logf("  Provider: %s", p.Provider)
	}
}

// =============================================================================
// JobService
// =============================================================================

func TestIntegration_ListJobs(t *testing.T) {
	skipIfNotIntegration(t)
	client := getTestClient(t)
	defer client.Close()

	ctx := context.Background()
	jobs, err := client.Jobs().ListJobs(ctx)
	if err != nil {
		// Jobs API may require elevated permissions not available to the test API key
		t.Logf("ListJobs() not available: %v", err)
		t.Skip("Jobs API not available (may require elevated permissions)")
	}

	t.Logf("Found %d jobs", len(jobs))
	for _, job := range jobs {
		t.Logf("  Job: Name=%s", job.Name)
	}
}

// =============================================================================
// TaurusNetwork - Participants
// =============================================================================

func TestIntegration_ListTaurusNetworkParticipants(t *testing.T) {
	skipIfNotIntegration(t)
	client := getTestClient(t)
	defer client.Close()

	ctx := context.Background()
	result, err := client.TaurusNetwork().Participants().ListParticipants(ctx, nil)
	if err != nil {
		// TaurusNetwork may not be enabled in all environments
		t.Logf("TaurusNetwork participants not available: %v", err)
		t.Skip("TaurusNetwork not available")
	}

	t.Logf("Found %d TaurusNetwork participants", len(result.Participants))
	for _, p := range result.Participants {
		t.Logf("  Participant: ID=%s, Name=%s", p.ID, p.Name)
	}
}

// =============================================================================
// TaurusNetwork - Pledges
// =============================================================================

func TestIntegration_ListTaurusNetworkPledges(t *testing.T) {
	skipIfNotIntegration(t)
	client := getTestClient(t)
	defer client.Close()

	ctx := context.Background()
	pledges, _, err := client.TaurusNetwork().Pledges().ListPledges(ctx, nil)
	if err != nil {
		// TaurusNetwork may not be enabled in all environments
		t.Logf("TaurusNetwork pledges not available: %v", err)
		t.Skip("TaurusNetwork not available")
	}

	t.Logf("Found %d TaurusNetwork pledges", len(pledges))
	for _, pledge := range pledges {
		t.Logf("  Pledge: ID=%s, Status=%s", pledge.ID, pledge.Status)
	}
}

// =============================================================================
// TaurusNetwork - Lending
// =============================================================================

func TestIntegration_ListTaurusNetworkLendingOffers(t *testing.T) {
	skipIfNotIntegration(t)
	client := getTestClient(t)
	defer client.Close()

	ctx := context.Background()
	result, err := client.TaurusNetwork().Lending().ListLendingOffers(ctx, nil)
	if err != nil {
		t.Logf("TaurusNetwork lending offers not available: %v", err)
		t.Skip("TaurusNetwork not available")
	}

	t.Logf("Found %d TaurusNetwork lending offers", len(result.LendingOffers))
}

func TestIntegration_ListTaurusNetworkLendingAgreements(t *testing.T) {
	skipIfNotIntegration(t)
	client := getTestClient(t)
	defer client.Close()

	ctx := context.Background()
	result, err := client.TaurusNetwork().Lending().ListLendingAgreements(ctx, nil)
	if err != nil {
		t.Logf("TaurusNetwork lending agreements not available: %v", err)
		t.Skip("TaurusNetwork not available")
	}

	t.Logf("Found %d TaurusNetwork lending agreements", len(result.LendingAgreements))
}

// =============================================================================
// TaurusNetwork - Settlements
// =============================================================================

func TestIntegration_ListTaurusNetworkSettlements(t *testing.T) {
	skipIfNotIntegration(t)
	client := getTestClient(t)
	defer client.Close()

	ctx := context.Background()
	result, err := client.TaurusNetwork().Settlements().ListSettlements(ctx, nil)
	if err != nil {
		t.Logf("TaurusNetwork settlements not available: %v", err)
		t.Skip("TaurusNetwork not available")
	}

	t.Logf("Found %d TaurusNetwork settlements", len(result.Settlements))
}

// =============================================================================
// TaurusNetwork - Sharing
// =============================================================================

func TestIntegration_ListTaurusNetworkSharedAddresses(t *testing.T) {
	skipIfNotIntegration(t)
	client := getTestClient(t)
	defer client.Close()

	ctx := context.Background()
	result, err := client.TaurusNetwork().Sharing().ListSharedAddresses(ctx, nil)
	if err != nil {
		t.Logf("TaurusNetwork shared addresses not available: %v", err)
		t.Skip("TaurusNetwork not available")
	}

	t.Logf("Found %d TaurusNetwork shared addresses", len(result.SharedAddresses))
}

func TestIntegration_ListTaurusNetworkSharedAssets(t *testing.T) {
	skipIfNotIntegration(t)
	client := getTestClient(t)
	defer client.Close()

	ctx := context.Background()
	result, err := client.TaurusNetwork().Sharing().ListSharedAssets(ctx, nil)
	if err != nil {
		t.Logf("TaurusNetwork shared assets not available: %v", err)
		t.Skip("TaurusNetwork not available")
	}

	t.Logf("Found %d TaurusNetwork shared assets", len(result.SharedAssets))
}

// =============================================================================
// ContractWhitelistingService
// =============================================================================

func TestIntegration_ListWhitelistedContracts(t *testing.T) {
	skipIfNotIntegration(t)
	client := getTestClient(t)
	defer client.Close()

	ctx := context.Background()
	result, err := client.WhitelistedContracts().ListWhitelistedContracts(ctx, nil)
	if err != nil {
		t.Fatalf("ListWhitelistedContracts() error = %v", err)
	}

	t.Logf("Found %d whitelisted contracts", len(result.Contracts))
}
